### DISCUSSION TITLE
response_model=Optional[ResponseType] leads to ResponseType as ref in openapi.json

### DISCUSSION THREAD

**Discussion by MarcoMuellner** [fastapi/fastapi#6254] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

import uvicorn
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}


@app.get("/someItem",
         response_model=Optional[str])
async def get_some_item():
    return None


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)  # noqa: S104, E261
```

### Description

* Open the browser at 0.0.0.0:8000/openapi.json
It returns (in part) the following json:

```json
{
  "openapi": "3.0.2",
  "info": {
    "title": "FastAPI",
    "version": "0.1.0"
  },
  "paths": {
    "/": {
      "get": {
        "summary": "Root",
        "operationId": "root__get",
        "responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {}
              }
            }
          }
        }
      }
    },
    "/someItem": {
      "get": {
        "summary": "Get Some Item",
        "operationId": "get_some_item_someItem_get",
        "responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response Get Some Item Someitem Get",
                  "type": "string"
                }
              }
            }
          }
        }
      }
    }
  }
}
```

However, the response type should be a oneOf instead of just the type. This is especially tedious when autogenerating clients from an openapi.json definition like this, as the client generator won't know that it can also expect None as a response value.

### Operating System

Linux, macOS

### FastAPI Version

0.78.0

### Python Version

3.9.6

**Comment by JarroVGIT** [comment, marked as answer]
There is no datatype of `null` in OpenAPI, so that would be impossible to implement I'd say. It works well for other types though, such as `response_model=Union[str, int]`, which does result in 

```json
"anyOf":[
    {
        "type":"string",
    },
    {
        "type":"integer",
    }
]
```

There is no type that would indicate the response could be None. 

EDIT: thinking about this, it would be pretty weird if your API gave two different response models for the same response status code. You either expect something or you don't, I would say it is pretty confusing that sometimes your endpoint returns a body, and sometimes it returns nothing. Even an empty object would be better than just nothing, I'd say.

Maybe something like this would be more clean:

```python
class NullableResponseModel(BaseModel):
    result: Union[str, None] = Field(None, nullable=True)


@app.get("/someItem", response_model=NullableResponseModel)
async def get_some_item():
    return None
```

The openapi.json would reflect that the `result` property of the response could be None (or, nullable):

```json
"/someItem": {
    "get": {
        "summary": "Get Some Item",
        "operationId": "get_some_item_someItem_get",
        "responses": {
            "200": {
                "description": "Successful Response",
                "content": {
                    "application/json": {
                        "schema": {
                            "$ref": "#/components/schemas/NullableResponseModel"
                        }
                    }
                }
            }
        }
    }
}
},
"components": {
"schemas": {
    "NullableResponseModel": {
        "title": "NullableResponseModel",
        "type": "object",
        "properties": {
            "result": {
                "title": "Result",
                "type": "string",
                "nullable": true
            }
        }
    }
}
```

**Comment by MarcoMuellner** [comment]
There are nullable types both in 3.0 and 3.1 (especially because 3.1 supports the full JSON schema)
https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-4.2.1

```
 null:  A JSON "null" value
```

This is a pretty smart workaround though

```python
@app.get("/someItem", response_model=NullableResponseModel)
async def get_some_item():
    return None
```

 by basically creating a new type for exactly what i need. But i don't feel that there is something inherently wrong in returning None or an object (so basically give me an object if it exists). And the `Optional[Type]` does describe also what the workaround does above.

Edit: 
It is also fully acceptable to have a anyOf or a oneOf return schema for Openapi 3.0, as prescribed here:
https://swagger.io/docs/specification/describing-responses/

and as you described above. I don't see why i can return a int or a string, but not a string or null

**Comment by JarroVGIT** [comment]
> There are nullable types both in 3.0 and 3.1 (especially because 3.1 supports the full JSON schema) https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-4.2.1

I didn't say it was a JSON issue, I said it is not a datatype in the OpenAPI spec.  See [here](https://swagger.io/docs/specification/data-models/data-types/).

```
Note that there is no null type; instead, the [nullable] attribute is used as a modifier of the base type.
```

Note that FastAPI uses the 3.0.2 spec of OpenAPI.

**Comment by MarcoMuellner** [comment]
Ah, i was not aware that FastAPI was on 3.0.2. Is this denoted in the doc somewhere, because i couldn't find it?

**Comment by JarroVGIT** [comment]
Not sure it is in the docs, but it is declared in the openapi.json file (at the top).

**Comment by MarcoMuellner** [comment]
Thanks for the help!
